ibis/sql/postgres/compiler.py

```diff
@@ -0,0 +1,519 @@
+        # blacklisted tokens
+        if curtokens:
+            reduced.append(sa.func.to_char(arg, ''.join(curtokens)))
+    return reduced
+
+
+def _strftime(t, expr):
+    arg, pattern = map(t.translate, expr.op().args)
+    tokens, _ = _scanner.scan(pattern.value)
+    reduced = _reduce_tokens(tokens, arg)
+    return reduce(sa.sql.ColumnElement.concat, reduced)
+
+
+def _distinct_from(a, b):
+    return ((a != None) | (b != None)) & sa.func.coalesce(a != b, True)
+
+
+def _find_in_set(t, expr):
+    # postgresql 9.5 has array_position, but the code below works on any
+    # version of postgres with generate_subscripts
+    # TODO: could make it even more generic by using generate_series
+    # TODO: this works with *any* type, not just strings. should the operation
+    #       itself also have this property?
+    arg, haystack = expr.op().args
+    needle = t.translate(arg)
+    haystack = sa.select([sa.literal(
+        [element._arg.value for element in haystack],
+        type_=sa.dialects.postgresql.ARRAY(needle.type)
+    ).label('haystack')]).cte()
+
+    subscripts = sa.select([
+        sa.func.generate_subscripts(haystack.c.haystack, 1).label('i')
+    ]).cte()
+
+    # return a zero based index
+    return sa.select([subscripts.c.i - 1]).where(
+        ~_distinct_from(haystack.c.haystack[subscripts.c.i], needle)
+    ).order_by(subscripts.c.i).limit(1)
+
+
+def _regex_replace(t, expr):
+    string, pattern, replacement = map(t.translate, expr.op().args)
+
+    # postgres defaults to replacing only the first occurrence
+    return sa.func.regexp_replace(string, pattern, replacement, 'g')
+
+
+def _variance_reduction(func_name):
+    suffix = {
+        'sample': 'samp',
+        'pop': 'pop'
+    }
+    def variance_compiler(t, expr):
+        arg, where, how = expr.op().args
+        func = getattr(sa.func, '%s_%s' % (func_name, suffix.get(how, 'samp')))
+        result = func(t.translate(arg))
+        return result if where is not None else sa.funcfilter(result, where)
+
+    return variance_compiler
+
+
+def _log(t, expr):
+    arg, base = expr.op().args
+    arg = t.translate(arg)
+    if base is not None:
+        return sa.func.log(t.translate(base), arg)
+    else:
+        return sa.func.ln(arg)
+
+
+_cumulative_to_reduction = {
+    ops.CumulativeSum: ops.Sum,
+    ops.CumulativeMin: ops.Min,
+    ops.CumulativeMax: ops.Max,
+    ops.CumulativeMean: ops.Mean,
+    ops.CumulativeAny: ops.Any,
+    ops.CumulativeAll: ops.All,
+}
+
+
+def _cumulative_to_window(translator, expr, window):
+    win = ibis.cumulative_window()
+    win = win.group_by(window._group_by).order_by(window._order_by)
+
+    op = expr.op()
+
+    klass = _cumulative_to_reduction[type(op)]
+    new_op = klass(*op.args)
+    new_expr = expr._factory(new_op, name=expr._name)
+
+    if type(new_op) in translator._rewrites:
+        new_expr = translator._rewrites[type(new_op)](new_expr)
+
+    return L.windowize_function(new_expr, win)
```

Comment:
This seems spurious. I'm not importing `L` anywhere, so I think this code doesn't get used. This should also go in the refactor of base functions.